Emissive system refactor

TemporalOroboros opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

The emissives system is the system that both lets computer screens and vendors glow in the dark and lets mobs and items block those glows. The current implementation relies on using filters to let mobs and items mask out the glow in the dark overlays on some structures. This is difficult to extend properly without massively increasing maptick. This PR changes the emissives system to use BYONDs native image layering to mask emissive overlays. This should prove to be a more extensible option.

tldr; There exists a system that lets computer screens glow on the dark and lets mobs and items block the glow. It isn't very extensible and this PR attempts to make it more extensible.

### How emissive overlays used to work

<details>
<summary>The details (This is a collapsing section. Click me.)</summary>

#### The current system

Currently emissive overlays and the emissive blockers that mask those overlays are handled using a system of inter-masking planes. The emissive overlays and the emissive blockers are placed on separate, hidden plane masters. These are essentially rendering layers and groups. The emissive blocker plane is then used to mask the emissive overlay plane which effectively allows the emissive blockers to block the emissive overlays from being seen. After is has been masked the emissive overlay plane is used to mask the lighting plane, essentially creating holes in the shadows wherever an unblocked glowing thing exists.

#### Why this is a potential problem

This system works fine. In fact it works great! The computer screens glow, any person or item that winds up on a computer blocks the glow, and everything just works. However, this system runs into some issues when you try to extend it to work on things other than structures. Namely, the current system only supports emissive overlays on structures and emissive overlays that are completely unblockable by any means. As a result, several interesting uses to the system require extending the system.

As a result, if you want to apply emissive overlays to items (which exist between structures and mobs) or emissive overlays to turfs (which exist below structures) you must extend the emissives system to get the emissive overlays and emissive blockers to properly function. Doing this naively, by adding extra emissive overlay and emissive blocker planes and applying all of the relevant masking filters, is not exactly performant.

For exactly how unperformant this system is, observe the following graph of maptick.

![](https://cdn.discordapp.com/attachments/326831214667235328/827816785016389642/maptick2.png)
Credit goes to TheFinalPotato/LemonInTheDark for the graph

Maptick is a major contributor to lag and the higher the maptick the more free lag you, the player, get delivered fresh to your client. See the massive spike? That is the result of #55782, an attempt to add glowing carpet to the game. Since the PR revolved around adding glowing carpet it had to extend the emissives system to allow for emissive turfs and emissive blocking structures. Extending the system was done naively as described above and you can see the results. 1.5 times the maptick across the board. Ouch.

So, we know that extending the system in it's current form is impractical. At least if done naively. Thus we are stuck.

</details>

tldr; The emissive system currently uses inter-plane masking to allow for emissive blockers to function. This is difficult to reasonably extend without murdering maptick. See #56496 for the results of naively extending this system.

### How emissive overlays are going to work

<details>
<summary>The details (This is a collapsing section. Click me.)</summary>

Alright, so we know that the current system of using planes to let the emissive blockers mask the emissive overlays is difficult to extend in it's current form. The solution is to change how the system works so that it can be extended in a more efficient manner. What we want is a system that allows one set of images to be out masked by another set of images and for the first set of images to be capable of masking the light plane. Preferably, we would also like the ability to interleave the masking effect between emissives and emissive blockers with almost arbitrary layering.

Conveniently, this layering and masking is something BYOND already does to normal items and objects. If we put the emissive overlays and the emissive blockers on the same plane we can use their layers to interleave them almost arbitrarily like any normal structures and items! All we need is a way to mask away the emissive blockers from the resulting rendered plane and we can mask the lighting plane with the remaining emissive overlays.

Luckily, BYOND has provided a single filter that is capable of this task. The color matrix filter. This filter can be used to apply a color matrix to an image! Provided that the emissive overlays and the emissive blockers are different colors we can use a color matrix filter to effectively mask out the emissive blockers from the plane! The resulting emissive plane can be applied as an alpha mask to the lighting plane as it used to, to the same effect. The best part is, we get layering practically for free!

This is exactly what this PR does. It converts the emissives system from the old plane and masking based blocking to a new layer-based system which uses BYONDs native layer handling to mask the emissive overlays.

</details>

tldr; This PR changes the emissives system to use BYONDs native image layering so extending it doesn't require additional plane masters.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
- Streamlined and significantly more extensible emissive overlays and emissive blockers.
- Possibly enables for the return of neon carpet without murdering maptick.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
refactor: Emissive overlays and blockers now use BYONDs native image layering.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

TemporalOroboros committed:
Layer-based emissives and emissive blockers

- Converts the emissive blocker system to work based off of layers.

ninjanomnom commented on code/__DEFINES/lighting.dm:
```diff
@@ -75,11 +75,31 @@
+GLOBAL_LIST_INIT(em_mask_matrix, EM_MASK_MATRIX)
+
+/// Produces a mutable appearance glued to the [EMISSIVE_PLANE] dyed to be a mutable appearance.
+/proc/emissive_appearance(icon, icon_state = "", layer = FLOAT_LAYER, alpha = 255, appearance_flags = NONE)
```
This is not a define

TemporalOroboros replied:
Helpers then?

ninjanomnom reviewed (changes requested):
Nice work on this, I've got quite a few things I'll want to use this for

TemporalOroboros committed:
Adds lighting helpers

- Moves the emissive appearance factory proc to lighting helpers at ninjanomnom's request.

MrPerson commented:
>Provided that the emissive overlays and the emissive blockers are different colors we can use a color matrix filter to effectively mask out the emissive blockers from the plane!

The 100% reliable way to know a great idea like this is that I feel like a moron for not thinking of it first.

LemonInTheDark commented:
How is this on client performance?

TemporalOroboros commented:
@LemonInTheDark 
> How is this on client performance?

I'm not sure currently. KylerAce said they were going to profile it when they can get their laptop.

LemonInTheDark commented:
Guess we could always give it a spin on live and see if anyone complains about client slowdown.
I'm not talking about maptick mind

TemporalOroboros commented:
Sounds like a plan.

LemonInTheDark commented:
Any news?

TemporalOroboros commented:
@LemonInTheDark

KylerAce says that it seems to be fine for maptick and I haven't heard anyone complaining about client performance. It should be good to go provided it doesn't conflict.

TemporalOroboros committed:
Merge branch 'master' into layer-emissives

TemporalOroboros committed:
Corrects fire alarm emissive overlay after master merge

TemporalOroboros committed:
Merge branch 'master' into layer-emissives

TemporalOroboros committed:
Merge branch 'master' into layer-emissives

TemporalOroboros committed:
Merge branch 'master' into layer-emissives
